File: icore/src/main/java/com/ddu/icore/ui/activity/BaseActivity.java
package com.ddu.icore.ui.activity;

import android.app.Application;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.ddu.icore.R;
import com.ddu.icore.aidl.GodIntent;
import com.ddu.icore.common.IObserver;
import com.ddu.icore.common.ObserverManager;
import com.ddu.icore.navigation.Navigator;
import com.ddu.icore.ui.view.TitleBar;
import com.ddu.icore.util.sys.ViewUtils;

public abstract class BaseActivity extends AppCompatActivity implements IObserver<GodIntent> {

    protected Context mContext;
    protected Application mApplication;
    protected Context mApplicationContext;

    protected ViewGroup mViewGroup;

    protected TitleBar mTitleBar;

    private OnFragmentCallback mOnFragmentCallback;

    protected LayoutInflater mLayoutInflater;

    protected View mCustomerTitleBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mContext = this;
        mApplication = getApplication();
        mApplicationContext = getApplicationContext();
        mLayoutInflater = LayoutInflater.from(mContext);
        registerObserver();
    }

    @Override
    public void setContentView(@LayoutRes int layoutResID) {
        if (isShowTitleBar()) {
            super.setContentView(R.layout.activity_base);
            mViewGroup = findView(R.id.ll_activity_base);
            mTitleBar = (TitleBar) mViewGroup.findViewById(R.id.ll_title_bar);
            getLayoutInflater().inflate(layoutResID, mViewGroup);
        } else {
            super.setContentView(layoutResID);
        }

    }

    public void startFragment(@NonNull Class fragment) {
        startFragment(fragment, null);
    }

    public void startFragment(@NonNull Class<? extends Fragment> fragment, Bundle bundle) {
        Navigator.startShowDetailActivity(this, fragment, bundle);
    }

    @NonNull
    public <T extends View> T findView(int resId) {
        return ViewUtils.findViewById(this, resId);
    }

    @Override
    public void registerObserver() {

    }

    @Override
    public void onReceiverNotify(GodIntent godIntent) {

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        ObserverManager.getInstance().unRegisterObserver(this);
    }

    public boolean isShowTitleBar() {
        return true;
    }

    public TitleBar getTitleBar() {
        return mTitleBar;
    }

    public void setDefaultTitle(int resId) {
        if (null != mTitleBar) {
            mTitleBar.setMiddleText(resId);
            setDefaultLeftImg();
        }
    }

    public void setDefaultTitle(String title) {
        if (null != mTitleBar) {
            mTitleBar.setMiddleText(title);
            setDefaultLeftImg();
        }
    }

    public void setRightText(String text, View.OnClickListener onClickListener) {
        if (null != mTitleBar) {
            TextView textView = mTitleBar.getRightText();
            textView.setText(text);
            textView.setOnClickListener(onClickListener);
        }
    }

    public void setRightImg(int resId, View.OnClickListener onClickListener) {
        if (null != mTitleBar) {
            ImageView imageView = mTitleBar.getRightImg();
            imageView.setImageResource(resId);
            imageView.setOnClickListener(onClickListener);
        }
    }

    private void setDefaultLeftImg() {
        if (null != mTitleBar) {
            mTitleBar.setDefaultLeftImg(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    onBackPressed();
                }
            });
        }
    }

    public interface OnFragmentCallback {
        boolean onFragmentKeyDownPress(int keyCode, KeyEvent keyEvent);

        boolean onFragmentBackPress();
    }

    public OnFragmentCallback getFragmentCallback() {
        return mOnFragmentCallback;
    }

    public void setFragmentCallback(OnFragmentCallback mOnFragmentCallback) {
        this.mOnFragmentCallback = mOnFragmentCallback;
    }
}

File: icore/src/main/java/com/ddu/icore/ui/activity/BaseActivity.kt
package com.ddu.icore.ui.activity

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.KeyEvent
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import com.ddu.icore.R
import com.ddu.icore.aidl.GodIntent
import com.ddu.icore.common.IObserver
import com.ddu.icore.common.ObserverManager
import com.ddu.icore.common.ext.find
import com.ddu.icore.navigation.Navigator
import com.ddu.icore.ui.widget.TitleBar

open class BaseActivity : AppCompatActivity(),
        IObserver {

    lateinit var mContext: Context

    lateinit var mViewGroup: ViewGroup

    var titleBar: TitleBar? = null
        protected set

    var fragmentCallback: OnFragmentCallback? = null

    protected var mCustomerTitleBar: View? = null

    private val appCompatDelegate: AppCompatDelegate? = null

    open fun isShowTitleBar() = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.setBackgroundDrawable(null)
        mContext = this
        registerObserver()

//        mImmersionBar = ImmersionBar.with(this)
//        mImmersionBar.statusBarDarkFont(true,0.2f).statusBarColor(R.color.c_ffffff).fitsSystemWindows(true).init()
    }

    override fun setContentView(@LayoutRes layoutResID: Int) {
        if (isShowTitleBar()) {
            super.setContentView(R.layout.i_activity_base)
            mViewGroup = find(R.id.ll_activity_base)
            titleBar = find(R.id.ll_title_bar)
            layoutInflater.inflate(layoutResID, mViewGroup)
        } else {
            super.setContentView(layoutResID)
        }

    }

    fun startFragment(fragment: Class<*>) {
        startFragment(fragment, null)
    }

    fun startFragment(fragmentName: String) {
        Navigator.startShowDetailActivity(this, fragmentName, null)
    }

    fun startFragment(fragmentName: String, bundle: Bundle?) {
        Navigator.startShowDetailActivity(this, fragmentName, bundle)
    }

    fun startFragment(fragment: Class<*>, bundle: Bundle?) {
        Navigator.startShowDetailActivity(this, fragment, bundle)
    }

    fun startActivity(activity: Class<out Activity>) {
        startActivity(Intent(this, activity))
    }

    open fun registerObserver() {
    }

    override fun onReceiverNotify(godIntent: GodIntent) {

    }

    override fun onDestroy() {
        super.onDestroy()
        ObserverManager.unRegisterObserver(this)
    }

    fun setDefaultTitle(resId: Int) {
        titleBar?.let {
            it.setMiddleText(resId)
            setDefaultLeftImg()
        }
    }

    fun setDefaultTitle(title: String) {
        if (null != titleBar) {
            titleBar!!.setMiddleText(title)
            setDefaultLeftImg()
        }
    }

    fun setRightText(text: String, onClickListener: View.OnClickListener) {
        titleBar?.setRightText(text, onClickListener)
    }

    fun setRightImg(resId: Int, onClickListener: View.OnClickListener) {
        if (null != titleBar) {
            val imageView = titleBar!!.rightImg
            imageView.setImageResource(resId)
            imageView.setOnClickListener(onClickListener)
        }
    }

    private fun setDefaultLeftImg() {
        if (null != titleBar) {
            titleBar!!.setDefaultLeftImg { onBackPressed() }
        }
    }

    interface OnFragmentCallback {
        fun onFragmentKeyDownPress(keyCode: Int, keyEvent: KeyEvent): Boolean

        fun onFragmentBackPress(): Boolean
    }

    override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
        return super.dispatchTouchEvent(ev)
    }
}
